### PR TITLE
[KEP-2400]: Add swap to kubectl describe node's output

### LINKED ISSUE #123962: [KEP-2400] Add ability to get node's swap capacity with `kubectl describe node`
### What would you like to be added?

Add ability to get node's swap capacity with `kubectl describe node`

### Why is this needed?

KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap

### PR DESCRIPTION
**Pull request by iholder101**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR provides a way of easily inspecting the amount of swap memory that's available on nodes.
This has been an ongoing request from the community for a long time now. Although this data already exists as part of `/stats/summary` and `/metrics/resource` (see https://github.com/kubernetes/kubernetes/pull/118865), using `kubectl describe node` is much simpler and straight forward.

##### Example output:
```bash
> kubectl describe node | grep Capacity -A15
Capacity:
  cpu:                40
  ephemeral-storage:  1293695812Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             394492620Ki
  pods:               110
  swap:               41943032Ki
Allocatable:
  cpu:                40
  ephemeral-storage:  1293695812Ki
  hugepages-1Gi:      0
  hugepages-2Mi:      0
  memory:             394492620Ki
  pods:               110
  swap:               41943032Ki
```

#### Which issue(s) this PR fixes:
Fixes #123962

#### Special notes for your reviewer:
I've added swap as a `ResourceName`. However, it's important to note that it **cannot** be added to the pod spec.

Another alternative is to use "swap" as a string and cast it to `ResourceName`, but then we'd need to keep it as a constant string somewhere and remember it is actually used as a ResourceName.

#### Does this PR introduce a user-facing change?

```release-note
Updated `kubectl describe node` to report the available amount of swap for Linux nodes.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap
```

### CODE REVIEW

**Comment by sftim** [comment]
(nit)
I recommend not putting backticks in commit messages (and, therefore, also not using them in PR titles)

**Comment by sftim** [comment]
Changelog suggestion
```diff
-It is now possible to get the amount of available swap on a node via `kubectl describe node`.
+Updated `kubectl describe node` to report the available amount of swap.
```

If this only applies to Linux nodes, consider stating that as well.

**Comment by sftim** [comment]
We could use `kubenetes.io/linux-swap` as a resource to report - see https://k8s.io/docs/tasks/administer-cluster/extended-resource-node/ for more about these.

**Comment by iholder101** [comment]
Hey @sftim! Thanks for having a look!

> I recommend not putting backticks in commit messages (and, therefore, also not using them in PR titles)

Done :+1: 

> Changelog suggestion
> If this only applies to Linux nodes, consider stating that as well.

Changed to:
```
Updated `kubectl describe node` to report the available amount of swap for Linux nodes.
```
WDYT?

> We could use kubenetes.io/linux-swap as a resource to report - see [k8s.io/docs/tasks/administer-cluster/extended-resource-node](https://k8s.io/docs/tasks/administer-cluster/extended-resource-node/) for more about these.

So IIUC the purpose of extended resources is to add a resource that k8s does not understand and that the admin needs to manage. However, since swap is a resource that's managed by kubelet and that Kubernetes does understand I thought that an extended resources won't be a perfect fit here.

Am I missing anything?

**Comment by sftim** [comment]
I can see how out of `swap` and `kubernetes.io/swap`, neither feels like the obvious right choice. I don't mind `swap`; I wanted to point out that we have an alternative if we want it.

**Comment by rphillips** [comment]
Swap would only be evaluated at kubelet start. Isn't the intention to have more swap added to the node after the kubelet starts?

**Comment by rphillips** [comment]
I am somewhat worried about exposing this via publicly available APIs would lead the impression that scheduler takes into account swap - which it doesn't.

**Comment by kannon92** [comment]
We added this based on feedback during KEP review process. https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963

The other issue is that users may not use swap as swap will be only used in cases of cgroup pressure or memory pressure.

**Comment by iholder101** [comment]
@rphillips 

> Swap would only be evaluated at kubelet start. Isn't the intention to have more swap added to the node after the kubelet starts?

While I agree it would be better if kubelet would be able to dynamically update the node's capacity, the same problem exists for other resources as well.

For example, it is possible to change the number of huge pages during runtime and the node's capabilities would not reflect that change.

Here's a quick experiment I did. First let's check the node's 2Mi hugepages count:
```bash
> k describe node node01 | grep hugepages-2Mi
  hugepages-2Mi:                  128Mi
```

On the node:
```bash
[node01 ~]$ sysctl vm.nr_hugepages
vm.nr_hugepages = 64
```

That makes sense, since 62*2Mi == 128Mi.
Now let's change the number of huge pages on the node:
```bash
[node01 ~]$ sysctl -w vm.nr_hugepages=128
vm.nr_hugepages = 128
[node01 ~]$ sysctl vm.nr_hugepages
vm.nr_hugepages = 128
```

But the node's capacity does not change:
```bash
> k describe node node01 | grep hugepages-2Mi
  hugepages-2Mi:                  128Mi
```

So to conclude, I think that while this is an unfortunate limitation, it is consistent to how Capacity reporting works with other resources.

**Comment by iholder101** [comment]
> I am somewhat worried about exposing this via publicly available APIs would lead the impression that scheduler takes into account swap - which it doesn't.

I understand where you're coming from.
However on the other hand, as opposed to other ResourceNames (memory, hugepages, etc) there is no way to explicitly ask for these resources. IOW, if a pod does not ask for swap, it's sensible to assume that swap capacity won't be a factor in its scheduling decisions.

Another point is that in the future we might want to add some API to explicitly ask for swap. If we'd choose to go down this path it might require that the scheduler would be aware of swap.

**Comment by liggitt** [comment]
I'm also surprised by making this visible to end users in the API

**Comment by iholder101** [comment]
> I'm also surprised by making this visible to end users in the API

Thanks for the feedback @liggitt.

The thing is that although we already expose swap capacity and per-container usage via `/stats/summary` and `/metrics/resource` endpoints [1], we've [received feedback](https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963) [2] regarding making it easier to debug and monitor swap usage / capacity.

I'm trying to find the right balance between making it easy to debug and monitor swap and not exposing too much information via APIs to ensure we're flexible enough to extend and change swap-related logic in the future, and avoid confusing users. If exposing this data via `kubectl describe node` and `kubectl top` is too much, I wonder which method should be used to expose this data.

@dchen1107, as you're the one who gave the initial feedback on this, would you be able to share your thoughts?

[1] https://github.com/kubernetes/kubernetes/pull/118865
[2] https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963

**Comment by yujuhong** [comment]
> I'm also surprised by making this visible to end users in the API

There's a trade-off adding this to the core API, and this one seems to be borderline. On one hand, we don't let users directly configure their pods to use it, so it seems reasonable to not introducing changes to the API. On the other hand, with limitedSwap enabled on the node, burstable QoS pods can use it, but it's hard to know/detect that from looking at the node or pod information without digging into the metrics. I think that's not a great user experience.

Has it been discussed in SIG-Node or previous KEP about how we expect the feature to be consumed? For example, do we expect the cluster admin to provision the nodes with swap and add labels/taints to them? I'm trying to understand whether this (the lack of visibility of the feature) is expected to be compensated by the provisioning operations?

**Comment by iholder101** [comment]
> > I'm also surprised by making this visible to end users in the API
> 
> There's a trade-off adding this to the core API, and this one seems to be borderline. On one hand, we don't let users directly configure their pods to use it, so it seems reasonable to not introducing changes to the API. On the other hand, with limitedSwap enabled on the node, burstable QoS pods can use it, but it's hard to know/detect that from looking at the node or pod information without digging into the metrics. I think that's not a great user experience.

Great summary!

> Has it been discussed in SIG-Node or previous KEP about how we expect the feature to be consumed? For example, do we expect the cluster admin to provision the nodes with swap and add labels/taints to them? I'm trying to understand whether this (the lack of visibility of the feature) is expected to be compensated by the provisioning operations?

So let me clarify what we currently have.

Firstly, swap information can be fetched from `/stats/summary` and `/metrics/resource` endpoints. This way users can inspect which nodes have / don't have swap, the swap capacity, and the swap usage on a pod/container level (see https://github.com/kubernetes/kubernetes/pull/118865 for more info).

In addition, we've added NFD support for swap here: https://github.com/kubernetes-sigs/node-feature-discovery/pull/1585.

So in summary, we do provide a way to understand which nodes support swap, with which capacity, and container/pod swap usage. However, I agree that it's not the best UX and the information is pretty hard to get. @dchen1107 specifically asked to expose this information with `kubectl describe node` (see https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963) and this PR implements that, but I'm happy to bring this up for discussion.

let me ping @dchen1107 again to open the door for further discussion.

**Comment by sftim** [comment]
Also have a think about how this might play with tooling like `swapd` (automatically managing the page file based on demand for it). The amount of paging storage available isn't fixed and can change in real time.

**Comment by dchen1107** [comment]
> > I'm also surprised by making this visible to end users in the API
> 
> Thanks for the feedback @liggitt.
> 
> The thing is that although we already expose swap capacity and per-container usage via `/stats/summary` and `/metrics/resource` endpoints [1], we've [received feedback](https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963) [2] regarding making it easier to debug and monitor swap usage / capacity.
> 
> I'm trying to find the right balance between making it easy to debug and monitor swap and not exposing too much information via APIs to ensure we're flexible enough to extend and change swap-related logic in the future, and avoid confusing users. If exposing this data via `kubectl describe node` and `kubectl top` is too much, I wonder which method should be used to expose this data.
> 
> @dchen1107, as you're the one who gave the initial feedback on this, would you be able to share your thoughts?
> 
> [1] #118865 [2] [kubernetes/enhancements#4401 (comment)](https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963)

It is hard to debug since there is no proper API to enable it for a node, but enabling swap might potentially affect performance, especially if the system frequently relies on it. There are two ways potentially exposing this via `kubectl describe node`:

1) Adding another Condition, such as SwapEnabled, similar to DiskPressure, PidPressure, etc. Or
2) Adding as an additional System Info, along with kernel version etc. @liggitt @yujuhong WDYT?

**Comment by liggitt** [comment]
Exposing existence of swap as a node condition for debugging seems the most reasonable of the options to me. No API change required, matches the approach taken for exposing arbitrary other node info by things like node problem detector, etc

**Comment by iholder101** [comment]
Thanks @dchen1107 and @liggitt.
I've created https://github.com/kubernetes/kubernetes/pull/125278 as an alternative to this PR. PTAL

**Comment by iholder101** [comment]
Closing this for now. Please see the discussion on https://github.com/kubernetes/kubernetes/pull/125278 for more details.

/close
